pyomo/solvers/tests/checks/test_GAMS.py

```diff
@@ -162,5 +165,146 @@ def test_keepfiles_gms(self):
+    using the gams command directly.
+
+    """
+
+    def test_no_tee(self):
+        with SolverFactory("gams", solver_io="gms") as opt:
+            with redirected_subprocess_call() as output:
+                opt.solve(self.m, tee=False)
+        self.assertFalse(output.getvalue())
+        self._check_logfile(exists=False)
+
+    def test_tee(self):
+        with SolverFactory("gams", solver_io="gms") as opt:
+            with redirected_subprocess_call() as output:
+                opt.solve(self.m, tee=True)
+        self.assertTrue(output.getvalue())
+        self._check_logfile(exists=False)
+
+    def test_logfile(self):
+        with SolverFactory("gams", solver_io="gms") as opt:
+            with redirected_subprocess_call() as output:
+                opt.solve(self.m, logfile=self.logfile)
+        self.assertFalse(output.getvalue())
+        self._check_logfile(exists=True)
+
+    def test_tee_and_logfile(self):
+        with SolverFactory("gams", solver_io="gms") as opt:
+            with redirected_subprocess_call() as output:
+                opt.solve(self.m, logfile=self.logfile, tee=True)
+        self.assertTrue(output.getvalue())
+        self._check_logfile(exists=True, expected=output.getvalue())
+
+
+@unittest.skipIf(not gamspy_available, "The 'gams' python bindings are not available")
+class GAMSLogfilePyTests(GAMSLogfileTestBase):
+    """Test class for testing permultations of tee and logfile options.
+
+    The tests build a simple model and solve it using the different options
+    using the python gams bindings.
+
+    """
+
+    def test_no_tee(self):
+        with SolverFactory("gams", solver_io="python") as opt:
+            with redirected_stdout() as output:
+                opt.solve(self.m, tee=False)
+        self.assertFalse(output.getvalue())
+        self._check_logfile(exists=False)
+
+    def test_tee(self):
+        with SolverFactory("gams", solver_io="python") as opt:
+            with redirected_stdout() as output:
+                opt.solve(self.m, tee=True)
+        self.assertTrue(output.getvalue())
+        self._check_logfile(exists=False)
+
+    def test_logfile(self):
+        with SolverFactory("gams", solver_io="python") as opt:
+            with redirected_stdout() as output:
+                opt.solve(self.m, logfile=self.logfile)
+        self._check_logfile(exists=True)
+        self.assertFalse(output.getvalue())
+
+    def test_tee_and_logfile(self):
+        with SolverFactory("gams", solver_io="python") as opt:
+            with redirected_stdout() as output:
+                opt.solve(self.m, logfile=self.logfile, tee=True)
+        self.assertTrue(output.getvalue())
+        self._check_logfile(exists=True, expected=output.getvalue())
+
+
+@contextlib.contextmanager
+def redirected_stdout():
+    """Temporarily redirect stdout into a string buffer."""
+    output = StringIO()
+    try:
+        setup_redirect(output)
+        yield output
+    finally:
+        reset_redirect()
```

Comment:
I like that this is implemented as a context manager.  Given that, the try-except is unnecessary.  We should be fine with:
```
def redirected_stdout():
    """Temporarily redirect stdout into a string buffer."""
    output = StringIO()
    setup_redirect(output)
    yield output
    reset_redirect()
```
This is so generally useful that we should promote it to pyomo.misc.redirect_io.

Reply:
I stand corrected.  I am a bit surprised with this, as I thought the point of a context manager was to ensure that the `__exit__` is always called.  Given this behavior, I am actually a bit surprised that the `contextlib.contextmanager` decorator actually works...